### PR TITLE
Gateway to router

### PR DESCRIPTION
**Pull request by cseed**
Proxy from gateway to router.  Logic for proxying web sockets came from here: https://stackoverflow.com/a/15198581/431282.

The Let's Encrypt stuff here isn't used anymore, not since we put letsencrypt in its own subproject.

I tested this by hand and then live on the cluster (!) and it's working fine.

### CODE REVIEW

File: gateway/Makefile [hunk 1]
```diff
@@ -31,10 +15,12 @@ deployment.yaml: deployment.yaml.in build
 	  -e "s,@image@,$(GATEWAY_IMAGE)," \
```
[the remark this answers is not in the document]
Comment (cseed) [reply gateway/Makefile]: Oops, I had renamed it to Dockerfile (doesn't need to be templated anymore) but it wasn't checked in.  Fixed.

**Comment by cseed** [comment]
I pushed a few more changes:

I added a nginx server block to handle proxying connections to other namespaces.  To map a namespace to the router service IP in gateway, I wrote a new server (router-resolver) that looks up the IP via kubernetes and call it with nginx auth_request.  So now you can hit, e.g., blah.cseed.internal.hail.is and it gets forwarded to the cseed:router service.

File: gateway/gateway.nginx.conf [hunk 2]
```diff
@@ -0,0 +1,71 @@
...
+        proxy_set_header Host $host;
+    }
+
+    location / {
+        auth_request /auth;
+        auth_request_set $router_ip $upstream_http_x_router_ip;
+        proxy_pass http://$router_ip$request_uri;
+
+        proxy_set_header Host $host;
+        proxy_http_version 1.1;
```
Comment (akotlar) [inline gateway/gateway.nginx.conf]: We need to set 

```
proxy_set_header X-Forwarded-Proto $scheme;
```

Comment (akotlar) [reply gateway/gateway.nginx.conf]: Else we cannot read the correct protocol from within the proxied service's server. This breaks notebook2, which relies on a callback url issued to auth0.

**Pull request merged**